bin/liquidator/src/liquidation_state.rs

```diff
@@ -0,0 +1,238 @@
+            }
+        } else {
+            error_tracking
+                .write()
+                .unwrap()
+                .clear(LiqErrorType::Liq, pubkey);
+        }
+    }
+
+    pub async fn maybe_liquidate_and_log_error(&mut self, pubkey: &Pubkey) -> anyhow::Result<bool> {
+        if self.should_skip_execution(pubkey) {
+            return Ok(false);
+        }
+
+        let result = liquidate::maybe_liquidate_account(
+            &self.mango_client,
+            &self.account_fetcher,
+            pubkey,
+            &self.liquidation_config,
+        )
+        .await;
+
+        self.log_or_ignore_error(&result, pubkey);
+        return result;
+    }
+}
+
+pub fn spawn_liquidation_job(
+    cli: &Cli,
+    shared_state: &Arc<RwLock<SharedState>>,
+    tx_trigger_sender: async_channel::Sender<()>,
+    mut liquidation: Box<LiquidationState>,
+    metrics: &Metrics,
+) -> JoinHandle<()> {
+    tokio::spawn({
+        let mut interval =
+            mango_v4_client::delay_interval(Duration::from_millis(cli.check_interval_ms));
+        let mut metric_liquidation_check = metrics.register_latency("liquidation_check".into());
+        let mut metric_liquidation_start_end =
+            metrics.register_latency("liquidation_start_end".into());
+
+        let mut liquidation_start_time = None;
+
+        let shared_state = shared_state.clone();
+        async move {
+            loop {
+                interval.tick().await;
+
+                let account_addresses = {
+                    let mut state = shared_state.write().unwrap();
+                    if !state.one_snapshot_done {
+                        // discard first latency info as it will skew data too much
+                        state.oldest_chain_event_reception_time = None;
+                        continue;
+                    }
+                    if state.oldest_chain_event_reception_time.is_none()
+                        && liquidation_start_time.is_none()
+                    {
+                        // no new update, skip computing
+                        continue;
+                    }
+
+                    state.mango_accounts.iter().cloned().collect_vec()
+                };
+
+                liquidation.errors.write().unwrap().update();
+                liquidation.oracle_errors.write().unwrap().update();
+
+                if liquidation_start_time.is_none() {
+                    liquidation_start_time = Some(Instant::now());
+                }
+
+                let found_candidates = liquidation
+                    .find_candidates(account_addresses.iter(), |p| {
+                        if shared_state
+                            .write()
+                            .unwrap()
+                            .liquidation_candidates_accounts
+                            .insert(p)
+                        {
+                            return tx_trigger_sender.try_send(()).map_err(|e| e.into());
```

Comment:
probably could write it as `tx_trigger_sender.try_send(())?;`. Could also go with `send()` since the channel is unbounded...